paasta_tools/paasta_remote_run.py

```diff
@@ -241,83 +334,210 @@ def task_config_to_dict(task_config):
+    runner = Sync(executor)
+    set_runner_signal_handlers(runner)
+    terminal_event = runner.run(task_config)
+    if getattr(terminal_event, 'platform_type', None) == 'lost':
+        runner.kill(task_config.task_id)
+    runner.stop()
+    return terminal_event
+
+
+def get_terminal_event_error_message(terminal_event):
+    if not terminal_event.terminal or terminal_event.success:
+        return None  # not terminal or not an error
+    elif TASKPROC_OFFER_TIMEOUT_RAW in terminal_event.raw:
+        return terminal_event.raw
+    else:
+        mesos_type = getattr(terminal_event, 'platform_type', None)
+        if mesos_type == 'failed':
+            error_message = ''
+        elif mesos_type == 'lost':
+            error_message = (
+                "- Task was lost probably due to a network partition or an "
+                "going away. It probably isn't coming back :("
+            )
+        elif mesos_type == 'error':
+            error_message = "- Encountered an unexpected error with Mesos"
+        else:
+            error_message = "- Unknown failure"
+
+        error_parts = [error_message] if error_message else []
+        error_parts.append(f"- Raw: {pprint.pformat(terminal_event.raw)}")
+        return '\n'.join(error_parts)
+
+
+def run_tasks_with_retries(executor_factory, task_config_factory, retries=0):
+    # use max in case retries is negative, +1 for initial try
+    tries_left = max(retries, 0) + 1
+    terminals = []
+
+    while tries_left > 0:
+        paasta_print(PaastaColors.yellow(
+            f"Scheduling task on Mesos (tries left: {tries_left})",
+        ))
+
+        try:
+            executor = executor_factory()
+            task_config = task_config_factory()
+            terminal_event = run_task(executor, task_config)
+        except (Exception, ValueError) as e:
+            # implies an error with our code, and not with mesos, so just return
+            # immediately
+            paasta_print(f"Except while running executor stack: {e}")
+            traceback.print_exc()
+            terminals.append((None, task_config))
+            return terminals
+
+        terminals.append((terminal_event, task_config))
+        if terminal_event.success:
+            paasta_print(PaastaColors.green("Task finished successfully"))
+            break
+        else:
+            # TODO: add reconciliation and other more specific behavior
+            error_msg = get_terminal_event_error_message(terminal_event)
+            paasta_print(PaastaColors.red(f"Task failed:\n{error_msg}"))
+
+        tries_left -= 1
+
+    return terminals
+
+
+def send_notification_email(
+    email_address,
+    framework_config,
+    task_config,
+    success=True,
+    error_message=None,
+):
+    success_str = 'succeeded' if success else 'failed'
+
+    msg = EmailMessage()
+    msg['From'] = email_address
+    msg['To'] = email_address
+    msg['Subject'] = (f"remote-run {success_str.upper()} - {task_config['name']}")
```

Comment:
Hm, should this be the run ID instead of the task_config name? I think we report the run ID when the task starts, not the task name.